torch/quantization/convert_modules.py

```diff
@@ -0,0 +1,221 @@
+
+    """
+    mod.add_module('quant', AbstractQuant(qConfig))
+    mod.add_module('dequant', AbstractDeQuant())
+    mod.forward = addQuantDeQuantCall(mod.forward)
+
+def get_config_key(name, qConfigDict):
+    if name in qConfigDict:
+        return name
+    elif name:
+        parent = '.'.join(name.split('.')[:-1])
+        return get_config_key(parent, qConfigDict)
+    else:
+        return '' if '' in qConfigDict else None
+
+def getModule(modInstance, name):
+    splits = name.split('.')
+    mod = modInstance._modules[splits[0]]
+    for i in range(1, len(splits)):
+        mod = mod._modules[splits[i]]
+    return mod
+
+def addQuantDeQuant(modInstance, qConfigDict):
+    r"""Add AbstractQuant and AbstractDeQuant module and modify forward function
+    to call quant dequant accordign to qConfigDict
+    """
+    mod_key_list = []
+    for name, mod in modInstance.named_modules():
+        dict_key = get_config_key(name, qConfigDict)
+        if dict_key is not None:
+            mod_key_list.append(dict_key)
+
+    for name in set(mod_key_list):
+        if name is not '':
+            addQuantDeQuantModule(getModule(modInstance, name), qConfigDict[name])
+        else:
+            addQuantDeQuantModule(modInstance, qConfigDict[name])
+
+def prepare(modInstance, qConfigDict):
+    r"""Transform a module to a quantized module according to qConfigDict,
+    it adds observer and abstract quant dequant modules to the module. and
+    changes the forward method to invoke them
+
+    Args:
+        modInstance: instance of the module we want to transform
+        qCofnigDict: dictionary that maps from name of submodule to quantization
+                     configuration
+    """
+    addObserver(modInstance, qConfigDict)
+    addQuantDeQuant(modInstance, qConfigDict)
+
+class AbstractQuant(nn.Module):
+    r"""Abstract quantize module which will be replace to actual Quantize module
+    in swapModule function
+    """
+    def __init__(self, qconfig):
+        super(AbstractQuant, self).__init__()
+        self.add_module('observer', qconfig.activation())
+
+    def forward(self, x):
+        self.observer(x)
+        return x
+
+class AbstractDeQuant(nn.Module):
+    r"""Abstract dequantize module which will be replace to actual DeQuantize module
+    in swapModule function
+    """
+    def __init__(self):
+        super(AbstractDeQuant, self).__init__()
+
+    def forward(self, x):
+        return x
+
+def calculateQParams(MyModule):
+    r""" Calculates Quantizer parameters for activation tensors based on observer statistics
+    Updates the qparams parameter for each module with the quantizer params.
+
+    Args:
+        MyModule: Model with observer stats
+    """
+    if hasattr(MyModule, 'observer'):
+        if isinstance(MyModule.observer, Observer):
+            # Simple symmetric quantization over entire observed range
+            MyModule.qparams = MyModule.observer.calculate_qparams()
```

Comment:
why do we need to assign it on module level and create more pollution? why calculate_qparams can't be called during the conversion process?

Reply:
yeah I think we can move these to the place when we need them.